### PR TITLE
Tr/committable comments

### PR DESCRIPTION
**Pull request by mrT23**
### **PR Type**
enhancement, bug_fix

___

### **Description**
- Added diff code generation for code suggestions in Bitbucket, Bitbucket Server, GitHub, and GitLab providers.
- Improved logging and error handling across providers.
- Implemented validation for committable comments within PR hunks in the GitHub provider.
- Enhanced comment formatting with collapsible sections in GitLab provider.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>bitbucket_provider.py</strong><dd><code>Enhance Bitbucket provider with diff code generation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/git_providers/bitbucket_provider.py

<li>Added diff code generation for Bitbucket code suggestions.<br> <li> Improved logging for publishing code suggestions.<br> <li> Removed verbosity level checks for logging.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1369/files#diff-3f47323db85f534fa282b2d48c6cf542f00a5e791b62379253ef3d3e8c6bb3b2">+28/-13</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>bitbucket_server_provider.py</strong><dd><code>Enhance Bitbucket Server provider with diff code generation</code></dd></summary>
<hr>

pr_agent/git_providers/bitbucket_server_provider.py

<li>Added diff code generation for Bitbucket Server code suggestions.<br> <li> Improved logging for publishing code suggestions.<br> <li> Removed verbosity level checks for logging.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1369/files#diff-c9ca96d14ab7a2935714944f8f377c4a9bb425efde19e66595bb58d33e9f5a40">+26/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>github_provider.py</strong><dd><code>Validate and enhance GitHub committable comments</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/git_providers/github_provider.py

<li>Added validation for committable comments within PR hunks.<br> <li> Implemented diff code generation for suggestions.<br> <li> Improved handling of suggestion validation.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1369/files#diff-28979ec713529d6f619bcba0b9e6a3002d1164d47c70ccfea7f15306618a1a11">+105/-1</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>gitlab_provider.py</strong><dd><code>Enhance GitLab provider with diff code and formatting</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/git_providers/gitlab_provider.py

<li>Added diff code generation for GitLab suggestions.<br> <li> Improved comment formatting with collapsible sections.<br> <li> Enhanced fallback comment creation for GitLab.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1369/files#diff-35acba7a2829983fcf3c3c6a69b135988227e503ccdfa94366d1b48fcbbb0a31">+13/-8</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Review by hussam789 — APPROVED** [review]

**Comment by mrT23** [comment]
## Generating PR code suggestions


Work in progress ...<br>

File: pr_agent/git_providers/bitbucket_provider.py [hunk 1]
```diff
@@ -72,24 +74,38 @@ def publish_code_suggestions(self, code_suggestions: list) -> bool:
...
             body = suggestion["body"]
+            original_suggestion = suggestion.get('original_suggestion', None)  # needed for diff code
+            if original_suggestion:
+                try:
+                    existing_code = original_suggestion['existing_code'].rstrip() + "\n"
+                    improved_code = original_suggestion['improved_code'].rstrip() + "\n"
+                    diff = difflib.unified_diff(existing_code.split('\n'),
+                                                improved_code.split('\n'), n=999)
+                    patch_orig = "\n".join(diff)
+                    patch = "\n".join(patch_orig.splitlines()[5:]).strip('\n')
```
Comment (mrT23) [inline pr_agent/git_providers/bitbucket_provider.py:85]: **Suggestion:** Handle the case where `patch_orig.splitlines()[5:]` might result in an index error if the diff does not contain enough lines. [possible issue, importance: 6]
```suggestion
                    patch = "\n".join(patch_orig.splitlines()[5:]) if len(patch_orig.splitlines()) > 5 else ""
```

File: pr_agent/git_providers/bitbucket_provider.py [hunk 2]
```diff
@@ -72,24 +74,38 @@
...
+                    diff_code = f"\n\n```diff\n{patch.rstrip()}\n```"
+                    # replace ```suggestion ... ``` with diff_code, using regex:
+                    body = re.sub(r'```suggestion.*?```', diff_code, body, flags=re.DOTALL)
```
Comment (mrT23) [inline pr_agent/git_providers/bitbucket_provider.py:88]: **Suggestion:** Replace the regex pattern in `re.sub` to ensure it correctly targets only the intended markdown fenced code blocks. [general, importance: 5]
```suggestion
                    body = re.sub(r'```suggestion[\s\S]*?```', diff_code, body, flags=re.DOTALL)
```

File: pr_agent/git_providers/github_provider.py [hunk 3]
```diff
@@ -872,3 +879,100 @@ def auto_approve(self) -> bool:
...
+                                d1_clip = abs(min(0, d1))
+                                d2_clip = abs(min(0, d2))
+                                d = max(d1_clip, d2_clip)
+                                if d < min_distance:
+                                    patch_range_min = patch_range
+                                    min_distance = min(min_distance, d)
+                        if not is_valid_hunk:
+                            if min_distance < 10:  # 10 lines - a reasonable distance to consider the comment inside the hunk
+                                # make the suggestion non-committable, yet multi line
+                                suggestion['relevant_lines_start'] = max(suggestion['relevant_lines_start'], patch_range_min['start'])
```
Comment (mrT23) [inline pr_agent/git_providers/github_provider.py:954]: **Suggestion:** Validate the `validate_comments_inside_hunks` function to ensure it correctly identifies and handles comments outside valid hunks. [possible issue, importance: 7]
```suggestion
                            if not is_valid_hunk:
                                if min_distance < 10 and patch_range_min:
                                    ...
```

**Pull request merged**